ci, iwyu: Treat warnings as errors for `src/init` and `src/policy`

hebasto opened the pull request:
This PR continues the work from bitcoin/bitcoin#31308 by extending the treatment of IWYU warnings as errors to the `src/init` and `src/policy` directories.

hebasto committed:
ci, iwyu: Fix warnings in `src/init` and treat them as errors

hebasto commented on src/policy/truc_policy.cpp:
```diff
@@ -4,14 +4,18 @@
+#include <txmempool.h>
 #include <util/check.h>
 
+#include <boost/multi_index/detail/hash_index_iterator.hpp>
```
> nit in [30e5b01](https://github.com/bitcoin/bitcoin/commit/30e5b015c973c1ac375e1eae6649238c6c2e2b28): This seems wrong, as there is no boost usage here?

It is used as in https://github.com/bitcoin/bitcoin/blob/a36b9320300d1fa1e5dbd1b4d493acf3578ac11a/src/policy/truc_policy.cpp#L218 the type of `mempool_ancestors.begin()` is `boost::multi_index::detail::hashed_index_iterator<...>`.

However, the `detail` namespace should be hidden with a Boost-specifc mapping file.

maflcko replied:
> It is used as in

I understand, but I wonder what value there is in listing the detailed internal boost includes. The `mempool_ancestors` already requires the `txmempool.h` include, which contains all required boost includes. No other module will have to include them to be able to compile.

To me those just seems like bloat for no benefit and the downside of making it harder to change a mempool implementation detail, because headers all over the codebase will have to be adjusted to please iwyu.

hebasto committed:
ci, iwyu: Add latest Boost mapping file

maflcko reviewed (approved):
review ACK 8ae41ae6f2c2ccc129a84f8f2cc8ae415e2125e0 🐇

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 8ae41ae6f2c2ccc129a84f8f2cc8ae415e2125e0 🐇
/yGQlCLoD+y92eefuzREel2J883stbBwa/+Pt1X3bzg5mC8jGDGvjG7pgVwg7TuvT7ZyPcl1EZc2Tk7H1ZdqDg==
```

</details>

hebasto commented on src/policy/policy.cpp:
```diff
@@ -18,9 +18,11 @@
 
 #include <algorithm>
 #include <cstddef>
+#include <span>
```
Right. But let me leave this change for a follow-up, as it would touch files in `src/crypto`.

maflcko replied:
Yeah, I can see it both ways. I think the possible harm from including span.h over span is limited with this trivial header. Just noting I did the same with the time.h include: https://github.com/bitcoin/bitcoin/blob/3bb30658e631ed45b6c8609292facc7ae3dd0f61/src/util/time.h#L9

Also for the validation.h include: https://github.com/bitcoin/bitcoin/blob/3bb30658e631ed45b6c8609292facc7ae3dd0f61/src/validation.h#L19. There, the harm could be larger, as the validation header is a bit larger than the cs_main header, but it feels a bit verbose to redundantly include cs_main where validation is included.

Though, no strong opinion. This is just my thinking and anything is fine here, as long as reviewers are happy with it.

hebasto committed:
ci, iwyu: Export `<span>` header from `<span.h>`

hebasto committed:
ci, iwyu: Fix warnings in `src/policy` and treat them as errors

Headers that are no longer available via transitive includes are now
included directly in `bitcoin-tx.cpp` and sources in `src/test`.

fanquake commented on ci/test/03_test_script.sh:
```diff
@@ -224,8 +224,10 @@ if [ "${RUN_TIDY}" = "true" ]; then
+             -p "${BASE_BUILD_DIR}" "${MAKEJOBS}" -- \
+             -Xiwyu --cxx17ns \
+             -Xiwyu --mapping_file="${BASE_ROOT_DIR}/contrib/devtools/iwyu/bitcoin.core.imp" \
+             -Xiwyu --mapping_file=/include-what-you-use/boost-1.75-all.imp \
```
> include-what-you-use/boost-1.75-all.imp

I guess this is still correct-enough to use here, but are we making any effort to usptream updates, given it's 14 Boost releases out of date (current release is 1.89.0)? Seems possible that this could lead to incorrect IWYU suggestions, given any recent Boost refactorings?

hebasto commented:
> > this pull request conflicts with the following ones:
> > #33629 (Cluster mempool by sdaftuar)
> > #33591 (Cluster mempool followups by sdaftuar)
> 
> I think if you want to continue these refactors, it might be better if you picked code that doesn't conflict with Cluster Mempool or Kernel work, or things we are trying to ship in `v31`.

I agree, and when I noticed these conflicts, I checked that those PRs don't have any ACKs yet.

> Could instead do directories that rarely change, like `qt`, or `zmq`?

Right.

maflcko commented:
Other candidates could be: `common/ compat/ primitives/ script/ support/ util/ zmq/`
